src/main/java/org/fusesource/jansi/internal/AnsiConsoleSupportHolder.java

```diff
@@ -81,47 +82,58 @@ private static AnsiConsoleSupport findProvider(String providerList) {
             err = e;
         }
 
-        String providerName = null;
-        AnsiConsoleSupport.CLibrary clib = null;
-        AnsiConsoleSupport.Kernel32 kernel32 = null;
-
-        if (ansiConsoleSupport != null) {
-            try {
-                providerName = ansiConsoleSupport.getProviderName();
-                clib = ansiConsoleSupport.getCLibrary();
-                kernel32 = OSInfo.isWindows() ? ansiConsoleSupport.getKernel32() : null;
-            } catch (Throwable e) {
-                err = e;
-            }
-        }
-
-        PROVIDER_NAME = providerName;
-        CLIBRARY = clib;
-        KERNEL32 = kernel32;
+        PROVIDER = ansiConsoleSupport;
         ERR = err;
     }
 
     public static String getProviderName() {
-        return PROVIDER_NAME;
+        return PROVIDER.getProviderName();
+    }
+
+    private static final class LibraryHolder {
+        static final AnsiConsoleSupport.CLibrary CLIBRARY;
+        static final AnsiConsoleSupport.Kernel32 KERNEL32;
+        static final Throwable ERR;
+
+        static {
+            AnsiConsoleSupport.CLibrary clib = null;
+            AnsiConsoleSupport.Kernel32 kernel32 = null;
+            Throwable err = null;
+
+            if (PROVIDER != null) {
+                try {
+                    clib = PROVIDER.getCLibrary();
+                    kernel32 = OSInfo.isWindows() ? PROVIDER.getKernel32() : null;
+                } catch (Throwable e) {
+                    err = e;
+                }
+            } else {
+                err = AnsiConsoleSupportHolder.ERR;
+            }
+
+            CLIBRARY = clib;
+            KERNEL32 = kernel32;
+            ERR = err;
+        }
     }
 
     public static AnsiConsoleSupport.CLibrary getCLibrary() {
-        if (CLIBRARY == null) {
-            throw new RuntimeException("Unable to get the instance of CLibrary", ERR);
+        if (LibraryHolder.CLIBRARY == null) {
+            throw new RuntimeException("Unable to get the instance of CLibrary", LibraryHolder.ERR);
         }
 
-        return CLIBRARY;
+        return LibraryHolder.CLIBRARY;
     }
 
     public static AnsiConsoleSupport.Kernel32 getKernel32() {
-        if (KERNEL32 == null) {
+        if (LibraryHolder.KERNEL32 == null) {
             if (OSInfo.isWindows()) {
                 throw new RuntimeException("Unable to get the instance of Kernel32", ERR);
```

Comment:
I think this should be `LibraryHolder.ERR`.

On a side note, I'm wondering if we should use singleton patterns (a single static field for the instance holding fields for the variables) for `AnsiConsoleSupportHolder` and `LibraryHolder` instead of using a bunch of static fields...

Reply:
> On a side note, I'm wondering if we should use singleton patterns (a single static field for the instance holding fields for the variables) for `AnsiConsoleSupportHolder` and `LibraryHolder` instead of using a bunch of static fields...

I converted AnsiConsoleSupport to an abstract class and cached CLibrary and Kernel32 in it.